Implement timeout support for RRefs

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38590 Implement timeout support for RRefs**

This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

rohan-varma committed:
Implement timeout support for RRefs

This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -8,6 +8,9 @@ namespace torch {
 namespace rpc {
 
+// An enum denoting common RPC errors to allow specific error handling for them.
+enum RPCErrorType { TIMEOUT = 0, UNKNOWN_ERROR = 10 };
```
Maybe it might make more sense for unknown error to be 0?

jjlilley commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -20,6 +21,27 @@ namespace torch {
+      RpcAgent::getCurrentRpcAgent()->getTimeoutErrorDescription();
+  timeoutErrors.emplace_back("failed intentionally");
+
+  auto err = std::string(fm.error()->what());
```
I like that we're trying to distinguish between failure causes here.
Though scanning strings for a vector of possible error cases feels like it might be a bit cumbersome, and unnecessarily widen the RpcAgent api.

As a potential [half-baked] alternate suggestion, if we need the scan strings from the future error, maybe we can establish a convention where we tag the error string?

e.g. maybe something like "RpcErr:code"? 
    "RpcErr:1 RPC ran for more than {} milliseconds and timed out"
    "Host {} is down (RpcErr:2)"
    "Host {} is overloaded (RpcErr:3)"  ?
    etc.

(The tag needn't be at the beginning,  we can just strstr for the first occuring place, and if there's a parsable int, use that. Otherwise, categorize as unknown).

   "Rpc

rohan-varma replied:
Sure, this sounds reasonable. I think we can take the error type from the enum when we are handling the error (since we know what the error is) and then format that into the error message string. Then in this function, instead of parsing for substrings, we could instead parse out the error code as you suggested.

rohan-varma replied:
@jjlilley I updated the diff so that common errors (when we can detect them in the RPC agents) are tagged with the function `makeRPCError`, and this parsing logic is updated to attempt to get the tagged error from the error message.

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -9,17 +9,48 @@
+}
+
+std::string makeRPCError(std::string rpcErrorStr, RPCErrorType errorType) {
+  return fmt::format(
```
interesting, is it now legit to use libfmt in pytorch code?

rohan-varma replied:
I think since https://github.com/pytorch/pytorch/pull/37241, it is supported in libtorch

jjlilley replied:
> I think since #37241, it is supported in libtorch

awesome, good to know

jjlilley commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -9,17 +9,48 @@
+  if (pos != std::string::npos) {
+    // Parse the RPCErrorType.
+    auto errIdx = pos + torch::distributed::rpc::kRPCErrorPrefix.size() + 1;
+    auto errStr = err.substr(errIdx, err.find(":", errIdx) - errIdx);
```
what if err.find(':", errIdx) is npos?

rohan-varma replied:
I assumed that all users of this API would pass in an error created with the other function `makeRPCErrorType` (so the find() would succeed), although this of course doesn't have to be true. In that case, I think we can return unknown_errror.

jjlilley commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -232,6 +232,18 @@ class TORCH_API RRef : public RRefInterface {
   }
 
+  // Indicate that the creation of this RRef on owner node has timed out.
+  inline void setTimedOut() {
```
does this need to be a public method?

rohan-varma replied:
Changed to be protected.

jjlilley commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -8,6 +8,9 @@ namespace torch {
 namespace rpc {
 
+// An enum denoting common RPC errors to allow specific error handling for them.
+enum RPCErrorType { UNKNOWN_ERROR = 0, TIMEOUT = 1, INTENTIONAL_FAILURE = 2 };
```
what is meant by an "intentional failure"?
(even a 1-line comment might help clarify... is this like a user error, or an injected failure?)

rohan-varma replied:
This is from the faulty PG agent, which can intentionally fail messages, so it is indeed an injected failure. Will add a comment to clarify

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -7,12 +7,13 @@
 #include <atomic>
 #include <thread>
 
+#include <fmt/format.h>
```
even if we use libfmt, does this need to be included in a public header?

rohan-varma replied:
Moved it to the cpp

jjlilley commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -146,8 +172,11 @@ IValue UserRRef::toHere() const {
+      timeoutSeconds);
 
+  // TODO: we should ideally be able to interrupt this blocking wait if we check
+  // isTimedOut() and it is true.
```
isn't the wait() going to be interrupted by the futureResponse being satisfied? or is the issue that the rref_context needs to handleError, and we simply haven't had the timedout set yet?

(we could probably interpret any thrown exception here if it made sense?)

rohan-varma replied:
The specific case that I was thinking of is something like, 
1) Say rpc.remote() call is completely dropped on the receiver, (or takes a long time to process)
2) Sender, after rpc.remote(), calls rref.to_here(), thus blocking 
3) We are notified that (1) failed (via a timeout), and mark that future with an error and call `setTimedOut()` on the rref. 
4) However since in (2) we have already started blocking on a (different) future, we should ideally be able to interrupt this with the information we have from (3)

Although, this isn't too big of a deal since the blocking future in this call would eventually time out as well.

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -364,6 +364,19 @@ def remote(to, func, args=None, kwargs=None):
+                                   ``to`` is not successfully processed on this
+                                   worker within this timeout, then the next time
+                                   there is an attempt to use the RRef (such as
+                                   to_here()), a timeout will be raised indicating
```
double quote

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -364,6 +364,19 @@ def remote(to, func, args=None, kwargs=None):
+                                   infinite timeout, i.e. a timeout error will
+                                   never be raised. If not provided, the default
+                                   value set during initialization or with
+                                   `_set_rpc_timeout` is used.
```
ditto

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -18,10 +18,14 @@ void confirmPendingUser(
-  RRefContext::handleException(futureMessage);
 }
 
 c10::intrusive_ptr<RRef> finishCreatingOwnerRRef(
```
Do we also need to check timeout error here, in case the `remote` is sent to self?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2626,6 +2649,122 @@ def test_no_faulty_messages(self):
+        if self.rank != 0:
+            return
+        # test the case where rpc.remote() message creation is completely dropped.
+        dst_rank = (self.rank + 1) % self.world_size
```
Let's also test for sending to self.

rohan-varma replied:
Good catch! The next version of this diff will have thes updates.

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -203,12 +203,22 @@ PyObject* rpc_init(PyObject* /* unused */) {
           .def(
               "to_here",
-              &PyRRef::toHere,
+              [](PyRRef& self, float rpcTimeoutSeconds) {
```
Unneeded, removed

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -224,6 +241,25 @@ void RRefContext::delAllUsers(std::chrono::milliseconds timeoutMillis) {
+    for (auto& rrefId : unforkedOwners) {
+      LOG(INFO) << "Removing unforked OwnerRRef with RRefId: " << rrefId;
+      auto iter = owners_.find(rrefId);
+      owners_.erase(iter);
```
What's our protocol for potentially deleting the intrusive_ptr<RRef> in owners_ with the lock held?
In `delForkOfOwner()` we maintain the intrusive pointer until after the lock is dropped, perhaps to avoid excess work with the lock held.

(There's a separate comment about whether an O(owners_) operation is needed here, but I guess for delAllUsers, there probably should be.

rohan-varma replied:
@jjlilley , it seems like the only place where the `deletedRRef` is used is to explicitly acquire GIL and reset it if it holds a py object. I'm assuming this is to properly cleanup the python objects, so maybe we need to do something similar here.  @mrshenli Do you have any thoughts on this?

rohan-varma replied:
The O(owners_) point is true, though this function only runs once during RPC shutdown (and also has a couple blocking waits and RPC calls), although, if we are concerned about the performance, there may be some things we can do to make this portion function non-blocking .

jjlilley commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -98,6 +101,16 @@ void FaultyProcessGroupAgent::enqueueSend(SendWork work) {
+  float msgDelay = getDelayForMessage(message.type());
+  if (msgDelay != 0) {
+    // Sleep for the specified delay for the message.
+    std::this_thread::sleep_for(std::chrono::milliseconds(
```
Just a thought: Maybe to make the sleep action to be slightly more realistic (i.e. only affect the specified message, vs pausing everything behind it as well), it might be worth spawning a detached thread?, e.g.

```
 std::thread thr([msgDelay, message, this]() {
   std::this_thread::sleep_for(std::chrono::milliseconds(...);
   ProcessGroupAgent::sendToSelf(std::move(message));
  });
  thr.detach();
```

The detached thread is sketchy in prod code, seems reasonable for test cases

rohan-varma replied:
Great idea! This will definitely make the delays more realistic.

rohan-varma replied:
I will plan to do this in a follow up PR.

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -204,11 +204,19 @@ PyObject* rpc_init(PyObject* /* unused */) {
                   owner, returns a reference to the local value.
+
+                  Arguments:
+                        timeout (Optional, float): Timeout for ``to_here``. If
```
`timeout` in `rpc/api.py` are written as `timeout (float, optional)`

mrshenli commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.h:
```diff
@@ -51,6 +51,8 @@ class FaultyProcessGroupAgent : public ProcessGroupAgent {
   // Overrides ProcessGroupAgent's enqueueSend to inject delays.
   void enqueueSend(SendWork work) override;
+  // Override ProcessGroupAgent's sendToSelf to inject delays.
+  void sendToSelf(Message&& message) override;
```
Any reason this needs to be public?

rohan-varma replied:
Good point, we can actually make both `sendToSelf` and `enqueueSend` private in this class.

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -18,22 +18,40 @@ void confirmPendingUser(
+        ctx.getOwnerRRef(rrefId, /* ensure created */ true)->constValue();
+    auto errorType = getRPCErrorType(futureMessage);
+    rref_ptr->handleError(errorType, futureMessage);
+    // OwnerRRefs do not have a forkId, so dont need to assert here.
```
dont -> don't

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -18,22 +18,40 @@ void confirmPendingUser(
+    // We expect to run this callback only after the OwnerRRef has been created,
+    // since this is only invoked when sending to self.
+    auto rref_ptr =
+        ctx.getOwnerRRef(rrefId, /* ensure created */ true)->constValue();
```
This means if the RRef creation failed, it will crash one RPC thread?

mrshenli replied:
Does it make sense if we add a `getIfPresent/Exist` API? Or change the second arg to `blocking` instead of `forceCreated`?

rohan-varma replied:
The only way this can crash is if the ownerRRef is somehow removed from the `owners_` map before we run this callback, and this callback is (currently) only run when creating an RRef on the self worker. In all codepaths, we call `addSelfAsFork` which puts the RRef into the `owners_` map. Although, possibly one case where this issue can happen is if we process a message causing us to run `delForkOfOwner`.

rohan-varma replied:
I think one thing we could do is remove this forceCreated argument, and then call `wait`/`constValue` in `finishCreatingOwnerRRef`.

rohan-varma replied:
Sounds good, will keep it the way it is.

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -301,24 +316,27 @@ PyRRef pyRemotePythonUdf(
+        [ownerRRefId = ownerRRef->rrefId()](const FutureMessage& fm) {
+          auto deletedRRef = callback::finishCreatingOwnerRRef(fm, ownerRRefId);
+          if (deletedRRef && deletedRRef->isPyObj()) {
+            py::gil_scoped_acquire ag;
```
This probably is no longer necessary. Not requesting changes in this PR. I added https://github.com/pytorch/pytorch/issues/39355 to track and will investigate after this PR is merged.

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -18,22 +18,40 @@ void confirmPendingUser(
     auto rr = RemoteRet::fromMessage(futureMessage.constValue());
     TORCH_INTERNAL_ASSERT(rr->forkId() == expectedForkId);
+  } else {
+    // Handle errors, such as timeouts, by invoking the error handler on the
```
Would be be correct if I assume this is best-effort error handling? As this callback is only run when receiving the response, but before that, the applications can still fork and use this UserRRef? If yes, let's add this to comment and docs accrodingly.

rohan-varma replied:
Yes, will update docs to indicate this.

rohan-varma replied:
@mrshenli I updated this comment to include a note. Do you have a suggestion on the best place in docs to add this note? Should it be in the docstrings for `rpc.remote`?

rohan-varma replied:
@mrshenli I decided to add it in a docstring as `warning` under def remote. Let me know if something else might be better.

mrshenli replied:
This LGTM! Thanks!

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -364,6 +364,19 @@ def remote(to, func, args=None, kwargs=None):
+                                   ``to`` is not successfully processed on this
+                                   worker within this timeout, then the next time
+                                   there is an attempt to use the RRef (such as
+                                   ``to_here()``), a timeout will be raised
```
Not requesting changes to this PR if this is not supported yet. What will happen for the following code?

```python
x = rpc.remote(to, my_func, args=(...), timeout=2)
x.to_here() # blocking here
# after 2 seconds, my_func timed out
```

Will `to_here()` raise an exception after two seconds?

rohan-varma replied:
No, since currently, `to_here()` does a blocking call on the future corresponding to the `to_here` message, this would not happen. 

Currently, `to_here` would only notice `rpc.remote()` timed out if the callback corresponding to the remote call fired before the call to `to_here` and appropriately set `isTimedOut` which is checked in the beginning of `toHere`. Currently the user can pass a timeout to `to_here` and have that call time out. 

However, we should be able to support this by making `to_here()` wait on a CV notification that is notified when the future completes OR when `isTimedOut` is set. Then this should work as expected. I will look into this in a follow up diff.

rohan-varma replied:
Filed https://github.com/pytorch/pytorch/issues/39411 to track

mrshenli commented on torch/testing/_internal/distributed/rpc/jit/rpc_test_faulty.py:
```diff
@@ -23,7 +24,6 @@ def two_args_two_kwargs(
 ):
     return first_arg + second_arg + first_kwarg + second_kwarg
 
-
```
why removing this line?

rohan-varma replied:
Undid this change

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2582,6 +2583,28 @@ def test_callback_none(self):
+        with self.assertRaisesRegex(RuntimeError, "RRef creation"):
+            rref.to_here()
+
+        wait_until_owners_and_forks_on_rank(1, 1, rank=1)
```
does this mean applications also need to call this to wait for deletion?

rohan-varma replied:
No, this isn't expected to be called by users, it is for tests where we force specific errors (i.e. a timeout error on the client, but the server eventually processes the OwnerRRef creation, for example). Waiting in this way in tests will prevent flakiness/timeouts. In general, the user application won't really have a way to distinguish whether a message was completely dropped, or has timed out but will be completed at some point in the future. 

If the latter case is true in the user application, then there is a chance that RPC on the client node will shut down, and in this case the `RRefUserDeletes` sent out by the client won't have any effect. And if the server node shuts down after (in particular runs `_delete_all_user_rrefs`) it will incur a blocking timed wait on the owners and forks map in rref_context. This won't throw errors, but will delay RPC shutdown. I think if this becomes an issue, we can expose knobs to the user to control how long to wait on these messages during shutdown.

mrshenli reviewed (commented):
Thanks for adding a super rich set of tests!

rohan-varma commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -224,6 +241,25 @@ void RRefContext::delAllUsers(std::chrono::milliseconds timeoutMillis) {
+  // If an rref in the owners_ map has never been forked, we will never get a
+  // corresponding message from the forking node(s) telling us to delete the
+  // RRef. Hence we delete the RRef here. This can occur when a remote call is
+  // sent to self and times out.
```
@mrshenli , since the shutdown sequence is currently:
_delete_all_user_rrefs() (which calls this function)
_destroy_rref_context() (which also clears the owners map)

Since we run `delAllUsers` first, this deletion is here to avoid the long CV-based wait that waits for the owners map to be cleared. If we didn't have this and had the wait, then unittests could be flaky and time out (since if the condition described in the comment happens, owners_ size will be > 0).

mrshenli replied:
I see. Thanks for explaining.

rohan-varma commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -224,6 +241,25 @@ void RRefContext::delAllUsers(std::chrono::milliseconds timeoutMillis) {
+  // corresponding message from the forking node(s) telling us to delete the
+  // RRef. Hence we delete the RRef here. This can occur when a remote call is
+  // sent to self and times out.
+  {
```
Sure, I think we can call it `delAllUsersandUnforkedOwners`

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -383,6 +396,18 @@ def remote(to, func, args=None, kwargs=None):
+        ``remote`` fail, such as with a timeout error, we take a best-effort
+        approach to error handling. This means that errors are handled and set
+        on the resulting RRef on an asynchronous basis. If the RRef has not been
+        used by the application before this handling (such as to_here or
```
double quote on `to_here`

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -340,6 +391,17 @@ std::shared_ptr<Future<c10::intrusive_ptr<OwnerRRef>>> RRefContext::
     const c10::intrusive_ptr<RRef>& rref) {
+  // If we know that rref creation on the owner has timed out, raise it to the
+  // user here, otherwise continue with pickling.
+  // if (rref->getTimedOut()) {
```
delete commented-out code?

mrshenli reviewed (approved):
LGTM! Thanks for adding this! 

Please wait for tests before landing.

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]

rohan-varma commented:
ONNX error is unrelated.

rohan-varma committed:
Update on "Implement timeout support for RRefs"


This PR implements timeout semantics for RRef for parity with rpc_sync and rpc_async. How it works:

- Timeout parameter is added to rpc.remote. If the rpc.remote call times out, note that the error won't be raised to the user in that call, as it is not blocking (similar to rpc_async). Instead, the timeout error will be raised the next time the RRef is used (either by pickling or to_here call).
- Error handling semantics are added to RRef to deal with the timeout errors. Previously, if there was an error creating the OwnerRRef, the callback on the local user would throw an error in a callback, resulting in an `std::terminate`. Instead of this, the error is now caught and surfaced to the user the next time the RRef is used. As part of this, we have added an `RPCErrorType` enum and defined RRef error handlers to handle the `RPCErrorrTypes` (currently just timeout and unknown)
- A timeout parameter is added to `to_here()` which gives the user control over the max amount of time it can block for.
- Before blocking, `to_here` checks if the RRef creation on the remote node has timed out, and if so, throws with that information.
- `ctx.prepareChildForFork()` which is called when the RRef is pickled (i.e. used as an arg over RPC) checks if the `rpc.remote()` call had timed out, and if so, raises that error to the user.
- Tests are added, primarily via delay injection.

Differential Revision: [D21588165](https://our.internmc.facebook.com/intern/diff/D21588165/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21588165/)!

[ghstack-poisoned]
